Created a numpy version of the concentration driven mode

JGBroadbent opened the pull request:
Same formatting as emissions driven mode and tests written

# Pull request

Please confirm that this pull request has done the following:

- [x] Tests added
- [ ] Documentation added (where applicable)
- [ ] Example added (either to an existing notebook or as a new notebook, where applicable)
- [ ] Description in ``CHANGELOG.rst`` added

## Adding to CHANGELOG.rst

Please add a single line in the changelog notes similar to one of the following:

```
- (`#XX <http://link-to-pr.com>`_) Added feature which does something
- (`#XX <http://link-to-pr.com>`_) Fixed bug identified in (`#XX <http://link-to-issue.com>`_)
```

JGBroadbent committed:
Created a numpy version of the concentration driven mode

Same formatting as emissions driven mode and tests written

znicholls commented on fair/2.0.0/concentration_driven.py:
```diff
@@ -0,0 +1,155 @@
+    :obj:`pd.DataFrame`
+        Results of the run
+    """
+    raise NotImplementedError
```
```suggestion
```

znicholls commented on fair/2.0.0/concentration_driven.py:
```diff
@@ -0,0 +1,155 @@
+                            ext_forcing = cfg['ext_forcing'],\
+                            timestep = cfg['timestep'])
+    res = unifiedtools.convert_numpy_output_to_df(res_numpy)
+    res = pd.DataFrame()
```
```suggestion
```

znicholls commented on fair/2.0.0/concentration_driven.py:
```diff
@@ -0,0 +1,155 @@
+        (Concentration, Radiative Forcing, Temperature and Alpha)
+        Values are in :obj:`np.ndarray` format, with the final index representing 'timestep'
+    """
+    raise NotImplementedError
```
```suggestion
```

znicholls commented on fair/2.0.0/tests/numpy_tests/concentration_driven_test.py:
```diff
@@ -0,0 +1,263 @@
+import concentration_driven
+
+def test_single_species():
+    #emissions are based on exemplar CO2 emissions
```
```suggestion
    #emissions are based on example CO2 concentrations
```

??

JGBroadbent replied:
I did these tests by just altering my emissions tests! Should be conc!

znicholls commented on fair/2.0.0/tests/numpy_tests/concentration_driven_test.py:
```diff
@@ -0,0 +1,263 @@
+    assert np.allclose(RF_out, RF_compare)
+    assert np.allclose(alpha_out, alpha_compare)
+
+def test_zero_emissions():
```
```suggestion
def test_constant_concentrations():
```

JGBroadbent replied:
Fair enough

znicholls commented on fair/2.0.0/tools/unifiedtools.py:
```diff
@@ -81,4 +81,14 @@ def convert_numpy_output_to_df(res_numpy):
+
+def unstep_concentration(a,dt,alpha,tau,R_old, G_A):
+    
+    raise NotImplementedError
```
```suggestion
```

znicholls reviewed (commented):
Looks like a good start. I think this is the simplest way to add the extra run mode (even if some of the code feels duplicated, I can't see a clearer way to do it...). Do the tests pass (I'd be surprised given all the `NotImplementedError`s) or did you just want my review before you kept going?

JGBroadbent committed:
Update fair/2.0.0/tests/numpy_tests/concentration_driven_test.py

Co-authored-by: Zeb Nicholls <zebedee.nicholls@climate-energy-college.org>

JGBroadbent committed:
Update fair/2.0.0/tests/numpy_tests/concentration_driven_test.py

Co-authored-by: Zeb Nicholls <zebedee.nicholls@climate-energy-college.org>

JGBroadbent commented:
Hi Zeb,

I've not run any tests yet, just wanted to get your input before proceeding any further. RE the repeated code, I completely agree, there's a few repetitions of combinations of functions but mostly the whole thing is run out of common functions anyway.

JGBroadbent committed:
Apply suggestions from code review

Co-authored-by: Zeb Nicholls <zebedee.nicholls@climate-energy-college.org>

JGBroadbent committed:
Apply suggestions from code review

Co-authored-by: Zeb Nicholls <zebedee.nicholls@climate-energy-college.org>

JGBroadbent committed:
Apply suggestions from code review

Co-authored-by: Zeb Nicholls <zebedee.nicholls@climate-energy-college.org>

JGBroadbent committed:
Updated pandas df running

This version should now be able to do both forward and reverse runs, along with input/output in pandas dataframes.

JGBroadbent committed:
Reformated cfg

cfg dict reformatted to use pd.dataframes for gas, thermal params etc. Also built in support for aerosol emissions, through putting aer_conc into gas_params.

JGBroadbent committed:
Clean Up Emissions + Concentration Driven Modes and Add in Tests

Did some minor tweaks to try and be more pythonic and added tests in for pandas dataframe driven version of them.

znicholls commented on fair/2.0.0/tests/numpy_tests/emissions_driven_test.py:
```diff
@@ -261,3 +262,158 @@ def test_zero_emissions():
+                                2.245283e+106,\
+                                2.245283e+106]])
+
+  assert np.allclose(C_df.values, C_compare_np)
```
```suggestion
  np.testing.assert_allclose(C_df.values, C_compare_np)
```

This is the more standard pattern and means you get way more information if things do fail. Probably worth changing everything to match this pattern now.

JGBroadbent replied:
Done, will push commits with comments once all pushed through

znicholls commented on fair/2.0.0/tests/numpy_tests/emissions_driven_test.py:
```diff
@@ -261,3 +262,158 @@ def test_zero_emissions():
+  assert np.allclose(RF_df.values, RF_compare_np)
+  assert np.allclose(alpha_df.values, alpha_compare_np)
+
+  assert C_df.index.values == year_index_np
```
```suggestion
  np.testing.assert_array_equal(C_df.index.values, year_index_np)
```

Will give better error messages on failure

JGBroadbent replied:
Done, will push.

znicholls commented on fair/2.0.0/tests/numpy_tests/emissions_driven_test.py:
```diff
@@ -261,3 +262,158 @@ def test_zero_emissions():
+  assert RF_df.index.values == year_index_np
+  assert alpha_df.index.values == year_index_np
+
+  assert C_df.columns.values == gas_names
```
```suggestion
  assert C_df.columns.tolist() == gas_names
```

Just in case what's there fails, this might fix it

JGBroadbent replied:
Done, will push

znicholls reviewed (commented):
Looks good. Just to make sure I understand, the idea behind `convert_df_to_numpy` is that we always sort things alphabetically, and this assumption should then propagate through to all other uses?

JGBroadbent commented:
Yep, so I wanted to make sure that the input into the 'numpy version' was always properly formatted, but also I think it's not very user friendly to require people to enter the parameters, gasses etc all in the right order (or indeed in time order), so it makes sense to just sort everything by column and by index before proceeding. RE the openSCM runner, my impression is that I can fairly easily "bolt" that on top of this? Once I've got this to a desired standard I'll write a first draft of that, then come back and test everything.

znicholls commented:
> Yep, so I wanted to make sure that the input into the 'numpy version' was always properly formatted, but also I think it's not very user friendly to require people to enter the parameters, gasses etc all in the right order (or indeed in time order), so it makes sense to just sort everything by column and by index before proceeding

Cool that works. As we try and add more stuff we'll get a better sense of what the best system is so let's not sweat too much now.

> RE the openSCM runner, my impression is that I can fairly easily "bolt" that on top of this?

Yep. Let's get all the tests passing here and we're happy with our pandas run interface, bolting on openscm runner should be pretty simple.

JGBroadbent commented:
Great, will jump on that.

JGBroadbent committed:
Update fair/2.0.0/concentration_driven.py

Co-authored-by: Zeb Nicholls <zebedee.nicholls@climate-energy-college.org>

JGBroadbent committed:
Debugged and Tested

moved tests out of folder - nothing major!

Update .gitignore

Debugging

Debugged

Now passes all tests as defined on 30/09/2020

Cleaned code a bit

Update .gitignore

Delete pyvenv.cfg

Remove pytest env

znicholls commented:
nice @JGBroadbent, my suggested next steps would be:

1. rebase and remove the `tests/pytest-env` folder which is now in the repo. We don't want to be including virtual environments (and the 3 000 files they contain) in our repo!
1. merge this
1. I'll make a follow up merge request which adds continuous integration to our setup here (so the tests get run by github on pull requests as a final check)
1. then you make merge requests which add extra functionality (with the tests making sure we don't regress i.e. break anything in the process of our upgrades)
1. rinse and repeat, including other tests to add docs, linting etc. (I'll do these but I'll be a bit slow so let's not hold everything up) in the process

JGBroadbent commented:
I thought I’d excluded the tests/pytest! oh well, will do!

> On 30 Sep 2020, at 10:15, Zeb Nicholls <notifications@github.com> wrote:
> 
> 
> nice @JGBroadbent <https://github.com/JGBroadbent>, my suggested next steps would be:
> 
> rebase and remove the tests/pytest-env folder which is now in the repo. We don't want to be including virtual environments in our repo!
> merge this
> I'll make a follow up merge request which adds continuous integration to our setup here (so the tests get run by github on pull requests as a final check)
> then you make merge requests which add extra functionality (with the tests making sure we don't regress i.e. break anything in the process of our upgrades)
> rinse and repeat, including other tests to add docs, linting etc. (I'll do these but I'll be a bit slow so let's not hold everything up) in the process
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/OMS-NetZero/FAIR/pull/86#issuecomment-701268712>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQR67PFTBL26SRSILXVVUQDSILZLRANCNFSM4RW3BGFA>.
>

JGBroadbent commented:
Done

> On 30 Sep 2020, at 10:16, John Broadbent <johngeoffreybroadbent@gmail.com> wrote:
> 
> I thought I’d excluded the tests/pytest! oh well, will do!
> 
>> On 30 Sep 2020, at 10:15, Zeb Nicholls <notifications@github.com <mailto:notifications@github.com>> wrote:
>> 
>> 
>> nice @JGBroadbent <https://github.com/JGBroadbent>, my suggested next steps would be:
>> 
>> rebase and remove the tests/pytest-env folder which is now in the repo. We don't want to be including virtual environments in our repo!
>> merge this
>> I'll make a follow up merge request which adds continuous integration to our setup here (so the tests get run by github on pull requests as a final check)
>> then you make merge requests which add extra functionality (with the tests making sure we don't regress i.e. break anything in the process of our upgrades)
>> rinse and repeat, including other tests to add docs, linting etc. (I'll do these but I'll be a bit slow so let's not hold everything up) in the process
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub <https://github.com/OMS-NetZero/FAIR/pull/86#issuecomment-701268712>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQR67PFTBL26SRSILXVVUQDSILZLRANCNFSM4RW3BGFA>.
>> 
>
